lib/i18n-transform.js

```diff
@@ -52,93 +52,172 @@ var _getLanguageFromI18n = function(i18nObj) {
-        if(!language){
+    _mergeSelectedLanguage = function (destination, language) {
+        if (!language) {
             return destination;
         }
-        return hoek.merge(destination, language);
+        return _.assign(destination, language);
     },
 
-    _getSelectedLanguage = function(source, i18nArray, languages){
+    _getSelectedLanguage = function (source, i18nArray, languages, fields) {
         // expects the language object from the request to be passed in:
         // [
         //   { code: 'en', region: 'US', quality: 1.0 },
         //   { code: 'en',               quality: 0.8 },
         //    ...
         // ]
-        if(!languages || languages.length === 0){
-            return  _extractPrimaryLanguage(source, i18nArray);
+        if (!languages || languages.length === 0) {
+            return _extractPrimaryLanguage(source, i18nArray);
         }
         else {
-            languages.sort(function(a, b){
+            languages.sort(function (a, b) {
                 return b.quality - a.quality;
             });
 
-            return _extractPreferredLanguage(source, i18nArray, languages);
+            return fields ? _extractPreferredLanguagePerField(source, i18nArray, languages, fields)
+                : _extractPreferredLanguage(source, i18nArray, languages);
         }
-    };
+    },
 
-module.exports.transform = function(source, languages){
-    var i18nInfo = _getI18n(source);
+    _transform = function (source, languages, fields) {
+        var i18nInfo = _getI18n(source);
 
-    if(!i18nInfo){
-        return source;
-    }
+        if (!i18nInfo) {
+            return source;
+        }
 
-    var selectedLanguage = _getSelectedLanguage(source, i18nInfo.value, languages);
-    if (!selectedLanguage){
-        return;
-    }
+        var selectedLanguage = _getSelectedLanguage(source, i18nInfo.value, languages, fields);
+        if (!selectedLanguage) {
+            return;
+        }
 
-    var result = _mergeSelectedLanguage(source, selectedLanguage);
-    delete result[i18nInfo.name];
-    return result;
-};
+        var result = _mergeSelectedLanguage(source, selectedLanguage);
+        delete result[i18nInfo.name];
+        return result;
+    },
 
-module.exports.transformDestination = function(source, destinationToAddLanguageTo, languages, callback){
-    var i18nInfo = _getI18n(source);
+    _transformDestination = function (source, destinationToAddLanguageTo, languages, callback, fields) {
+        var i18nInfo = _getI18n(source);
 
-    if(!i18nInfo){
-        return source;
-    }
+        if (!i18nInfo) {
+            return source;
+        }
+
+        var selectedLanguage = _getSelectedLanguage(source, i18nInfo.value, languages, fields);
 
-    var selectedLanguage = _getSelectedLanguage(source, i18nInfo.value, languages);
+        if (!selectedLanguage) {
+            if (callback) {
+                callback(new Error("Could not find desired language"));
```

Comment:
How about
`new Error("Could not find the desired language " + selectedLanguage)`
?